include/rmm/mr/device/fixed_size_memory_resource.hpp

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <rmm/mr/device/device_memory_resource.hpp>
+#include <rmm/mr/device/default_memory_resource.hpp>
+
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+
+#include <cuda_runtime_api.h>
+
+#include <list>
+#include <unordered_map>
+#include <cstddef>
+#include <utility>
+#include <algorithm>
+#include <cassert>
+
+// forward decl
+using cudaStream_t = struct CUstream_st*;
+
+namespace rmm {
+
+namespace mr {
+/**---------------------------------------------------------------------------*
+ * @brief Allocates fixed-size memory blocks.
+ * 
+ * Supports only allocations of size smaller than the configured block_size.
+ *---------------------------------------------------------------------------**/
+class fixed_size_memory_resource : public device_memory_resource {
+ public:
+
+  
+  static constexpr std::size_t default_block_size = 1<<20; // 1 MiB
+  static constexpr std::size_t default_heap_chunk_size = 128 * default_block_size;
+  static constexpr std::size_t allocation_alignment = 256;
+
+  explicit fixed_size_memory_resource(std::size_t block_size = default_block_size,
+                                      std::size_t heap_chunk_size = default_heap_chunk_size,
+                                      rmm::mr::device_memory_resource *heap_resource =
```

Comment:
`fixed_size_memory_resource` should be a template on the `Upstream` resource. This allows passing in a pointer to a derived `device_memory_resource` type directly. The advantage of this is eliminating an extra virtual function call dispatch. See https://github.com/rapidsai/rmm/pull/297/files#diff-c01f255982463da036f688b697606f25R38 for example.